src/macros/macrorecorder.cpp

```diff
@@ -0,0 +1,108 @@
+#include "macrorecorder.h"
+
+#include <QtConcurrentRun>
+
+#include "control/controlproxy.h"
+#include "preferences/configobject.h"
+
+// TODO(xerus) handle track eject while recording
+
+MacroRecorder::MacroRecorder()
+        : m_COToggleRecording(ConfigKey(kMacroRecordingKey, "recording_toggle")),
+          m_CORecStatus(ConfigKey(kMacroRecordingKey, "recording_status")),
+          m_activeChannel(nullptr),
+          m_macroRecordingState(MacroState::Disabled),
+          m_pStartRecordingTimer(new QTimer()),
+          m_recordedMacro() {
+    qCDebug(macros) << "MacroRecorder construct";
+
+    connect(&m_COToggleRecording,
+            &ControlPushButton::valueChanged,
+            this,
+            &MacroRecorder::slotToggleRecording);
+    connect(m_pStartRecordingTimer,
+            &QTimer::timeout,
+            this,
+            &MacroRecorder::pollRecordingStart);
+}
+
+void MacroRecorder::notifyCueJump(ChannelHandle& channel, double origin, double target) {
+    qCDebug(macros) << "Jump in channel" << channel.handle();
+    if (checkOrClaimRecording(channel)) {
+        m_recordedMacro.appendJump(origin, target);
+        qCDebug(macros) << "Recorded jump in channel" << channel.handle();
+        setState(MacroState::Armed);
+    }
+}
+
+bool MacroRecorder::checkOrClaimRecording(ChannelHandle& channel) {
+    if (m_activeChannel != nullptr) {
+        return m_activeChannel->handle() == channel.handle() && claimRecording();
+    } else if (claimRecording()) {
+        m_activeChannel = &channel;
+        qCDebug(macros) << "Claimed recording for channel" << channel.handle();
+        return true;
+    }
+    return false;
+}
+
+bool MacroRecorder::claimRecording() {
+    auto armed = MacroState::Armed;
+    return m_macroRecordingState.compare_exchange_weak(armed, MacroState::Recording);
+}
+
+void MacroRecorder::pollRecordingStart() {
+    qCDebug(macros) << "Polling for recording start";
+    if (getState() != MacroState::Disabled) {
+        if (getActiveChannel() == nullptr) {
+            return;
+        } else {
+            ControlProxy(ConfigKey(kMacroRecordingKey, "recording_status")).set(2);
+        }
+    }
+    m_pStartRecordingTimer->stop();
+}
+
+void MacroRecorder::startRecording() {
+    qCDebug(macros) << "MacroRecorder recording armed";
+    m_CORecStatus.set(1);
+    m_recordedMacro.clear();
+    setState(MacroState::Armed);
+    m_pStartRecordingTimer->start(300);
+}
+
+void MacroRecorder::stopRecording() {
+    qCDebug(macros) << "MacroRecorder recording stop";
+    m_CORecStatus.set(0);
+    auto armed = MacroState::Armed;
+    // TODO(xerus) add concurrency test
+    while (!m_macroRecordingState.compare_exchange_weak(armed, MacroState::Disabled))
```

Comment:
"It's weak in the sense that it returns false even if the value of the object is equal to expected."

Since I am using it in a loop, using strong is only more costly, or am I understanding something wrong?

Reply:
Since you use this atomic to coordinate access to shared, non-atomic memory you need a memory barrier for synchronization. Those threads might be executed on different cores with separate caches.